tests/components/samsungtv/test_media_player.py

```diff
@@ -55,247 +55,316 @@ class UnhandledResponse(Exception):
+        with patch(
+            "homeassistant.components.samsungtv.media_player.LOGGER.warning"
+        ) as mocked_warn:
             add_entities = mock.Mock()
-            setup_platform(self.hass, {}, add_entities, discovery_info=None)
+            await async_setup_platform(hass, {}, add_entities, discovery_info=None)
             mocked_warn.assert_called_once_with("Cannot determine device")
             add_entities.assert_not_called()
 
-    def test_update_on(self):
-        """Testing update tv on."""
-        self.device.update()
-        self.assertEqual(STATE_ON, self.device._state)
 
-    def test_update_off(self):
-        """Testing update tv off."""
-        _remote = mock.Mock()
-        _remote.control = mock.Mock(side_effect=OSError("Boom"))
-        self.device.get_remote = mock.Mock(return_value=_remote)
-        self.device.update()
-        assert STATE_OFF == self.device._state
-
-    def test_send_key(self):
-        """Test for send key."""
-        self.device.send_key("KEY_POWER")
-        self.assertEqual(STATE_ON, self.device._state)
-
-    def test_send_key_broken_pipe(self):
-        """Testing broken pipe Exception."""
-        _remote = mock.Mock()
-        _remote.control = mock.Mock(side_effect=BrokenPipeError("Boom"))
-        self.device.get_remote = mock.Mock(return_value=_remote)
-        self.device.send_key("HELLO")
-        self.assertIsNone(self.device._remote)
-        self.assertEqual(STATE_ON, self.device._state)
-
-    def test_send_key_connection_closed_retry_succeed(self):
-        """Test retry on connection closed."""
-        _remote = mock.Mock()
-        _remote.control = mock.Mock(
-            side_effect=[
-                self.device._exceptions_class.ConnectionClosed("Boom"),
-                mock.DEFAULT,
-            ]
-        )
-        self.device.get_remote = mock.Mock(return_value=_remote)
-        command = "HELLO"
-        self.device.send_key(command)
-        self.assertEqual(STATE_ON, self.device._state)
-        # verify that _remote.control() get called twice because of retry logic
-        expected = [mock.call(command), mock.call(command)]
-        assert expected == _remote.control.call_args_list
-
-    def test_send_key_unhandled_response(self):
-        """Testing unhandled response exception."""
-        _remote = mock.Mock()
-        _remote.control = mock.Mock(
-            side_effect=self.device._exceptions_class.UnhandledResponse("Boom")
-        )
-        self.device.get_remote = mock.Mock(return_value=_remote)
-        self.device.send_key("HELLO")
-        self.assertIsNone(self.device._remote)
-        self.assertEqual(STATE_ON, self.device._state)
+async def test_setup_method_selection():
+    """Test of method selection."""
+    conf = WORKING_CONFIG.copy()
+    device = SamsungTVDevice(**conf)
+    conf[CONF_PORT] = 8001
+    assert device._config[CONF_METHOD] == "websocket"
```

Comment:
We don't need to check internal attributes of the entity. Those are details that may change in a refactor. Changing those things are ok, as long as the entity behaves correctly by setting the correct state and providing the correct service methods.

Reply:
The only way to check if the method selection worked is to look at this or to check the `SUPPORT_SAMSUNGTV`. Should I switch to this?

Reply:
We should patch `samsungctl.Remote.control` and assert that it is called with the correct argument, ie the correct key, either `"KEY_POWER"` or `"KEY_POWEROFF"` depending on method, when calling the turn_off service targeting the added entity.